data-processing-lib/spark/Dockerfile

```diff
@@ -0,0 +1,89 @@
+
+# Add further jar files needed to run S3 in Kubernetes
+RUN wget https://repo1.maven.org/maven2/com/fasterxml/woodstox/woodstox-core/6.6.2/woodstox-core-6.6.2.jar && \
+    wget https://repo1.maven.org/maven2/org/codehaus/woodstox/stax2-api/4.2.2/stax2-api-4.2.2.jar && \
+    wget https://repo1.maven.org/maven2/org/apache/commons/commons-configuration2/2.10.1/commons-configuration2-2.10.1.jar && \
+    wget https://repo1.maven.org/maven2/com/google/re2j/re2j/1.7/re2j-1.7.jar
+
+RUN rm hadoop-shaded-guava-1.1.1.jar
+
+# use the latest python 3.11 slim bookworm image
+# ARG PYTHON_IMAGE=python:3.11.9-slim-bookworm
+# FROM $PYTHON_IMAGE
+FROM python:3.11.9-slim-bookworm
+
+# install additional packages required by spark
+RUN set -ex && \
+    apt-get update && \
+    ln -s /lib /lib64 && \
+    apt install -y bash tini libc6 libpam-modules krb5-user libnss3 procps net-tools && \
+    rm /bin/sh && \
+    ln -sv /bin/bash /bin/sh && \
+    echo "auth required pam_wheel.so use_uid" >> /etc/pam.d/su && \
+    chgrp root /etc/passwd && chmod ug+rw /etc/passwd && \
+    rm -rf /var/cache/apt/* && rm -rf /var/lib/apt/lists/*
+
+# Copy JRE from the Open JRE build image
+ENV JAVA_HOME=/opt/java/openjdk
+COPY --from=build $JAVA_HOME $JAVA_HOME
+# add java to executable path
+ENV PATH="${JAVA_HOME}/bin:${PATH}"
+
+# Set /opt/spark as SPARK_HOME and copy Spark files there from the build image
+ENV SPARK_HOME /opt/spark
+COPY --from=build ${SPARK_HOME}/jars ${SPARK_HOME}/jars
+COPY --from=build ${SPARK_HOME}/RELEASE ${SPARK_HOME}/RELEASE
+COPY --from=build ${SPARK_HOME}/bin ${SPARK_HOME}/bin
+COPY --from=build ${SPARK_HOME}/sbin ${SPARK_HOME}/sbin
+COPY --from=build ${SPARK_HOME}/kubernetes/dockerfiles/spark/entrypoint.sh /opt/
+COPY --from=build ${SPARK_HOME}/kubernetes/dockerfiles/spark/decom.sh /opt/
+COPY --from=build ${SPARK_HOME}/data ${SPARK_HOME}/data
+
+WORKDIR ${SPARK_HOME}/work-dir
+RUN chmod g+w ${SPARK_HOME}/work-dir
+RUN chmod a+x /opt/decom.sh
+
+# Install python packages (other than pyspark) using pip for pyspark to work.
+# Pyspark is installed in the COPY instruction below
+RUN mkdir ${SPARK_HOME}/python && \
+    pip3 install --upgrade pip setuptools && \
+    pip3 install --upgrade --trusted-host pypi.org --trusted-host files.pythonhosted.org pyyaml pyarrow boto3 && \
+    rm -rf /root/.cache && rm -rf /var/cache/apt/* && rm -rf /var/lib/apt/lists/*
+
+COPY --from=build ${SPARK_HOME}/python/pyspark ${SPARK_HOME}/python/pyspark
+COPY --from=build ${SPARK_HOME}/python/lib ${SPARK_HOME}/python/lib
+RUN echo "spark:x:185:0::/opt/spark/work-dir:/bin/bash" >> /etc/passwd
+WORKDIR ${SPARK_HOME}/work-dir
+ENTRYPOINT [ "/opt/entrypoint.sh" ]
+
+# Specify the User that the actual main process will run as
+ARG spark_uid=185
+USER ${spark_uid}
+
+# add pyspark, py4j to PYTHONPATH
+ENV PYTHONPATH=${SPARK_HOME}/python/lib/pyspark.zip:${SPARK_HOME}/python/lib/py4j-0.10.9.7-src.zip:${PYTHONPATH}
```

Comment:
1. We are already using `docker-image-tool.sh` tool to build the image.  Using this script directly is not a good practice, as it would require downloading a spark release on the local machine, and then running this script.  This approach cannot be automated, so we have encapsulated the download of the spark release and the invocation of this script in the docker image.

2. The bitnami spark image has several shortcomings - it is not built for running Spark in kubernetes, and it is not built for running Spark with S3.  As such, it requires several changes made to it, for example changing the `ENTRYPOINT`, without which it will not run in Kubernetes (on the driver node, or on the executor nodes). So we cannot use this image as is, and at this point I do not see any value in starting from this image, rather than from our base image.  Moreover, by looking at the size of this image (1.65GB), it is much larger than the size of our image (965MB); to use this image we will need to do additional investigation as to which packages are included in it, and where does the difference in size come from.